File: algo-c-to-_/examples/sqrt_.py
#
#	from src/sqrt.c
#
#	double mysqrt(double)		to	fSqrt
#
#	from src/isqrt.c
#
#	unsigned isqrt(unsigned)	to	iSqrt
#

from nthroot import fSqrt, iSqrt
from math import sqrt
from sys import exit

def _t_iSqrt(l, r):
    for i in range(l, r):
        t = iSqrt(i)
        if t * t > i or (t+1) * (t+1) <= i:
            print("ERROR: {0:d} {1:d}".format(i, t))
            exit()

    print("iSqrt() seems to be fine in {0:d} to {1:d}-1.".format(l, r))

for i in range(21):
    print("math.sqrt({0})\t{1:.20f}\nfSqrt({2})\t{3:.20f} ({4})".format(
        i, sqrt(i), i, fSqrt(i), sqrt(i) == fSqrt(i)))

print()

# _t_iSqrt(0, 1 << 32) # Maybe this will take a lot of time. (*)
_t_iSqrt(0, 1 << 16)

#
#	*) Please choose a number depending on the purpose and situation.
#
#	e.g. a list of elapsed time - running _t_iSqrt(0, n) on my old cheap laptop
#
#	in luajit, n == 1<<28 (2^28):
#
#		$ time LUA_PATH=src/?.luajit luajit example/sqrt.luajit
#		...
#		iSqrt() seems to be fine in 0 to 268435456-1.
#
#		real    0m36.307s
#		user    0m35.889s
#		sys     0m0.015s
#
#	in luajit, n == 1<<32 (2^32):
#
#		$ time LUA_PATH=src/?.luajit luajit example/sqrt.luajit
#		...
#		iSqrt() seems to be fine in 0 to 4294967296-1.
#
#		real    10m0.827s
#		user    10m0.906s
#		sys     0m0.096s
#
#	in lua, n == 1<<28 (2^28)
#
#		$ time LUA_PATH=src/?.lua lua example/sqrt.lua
#		...
#		iSqrt() seems to be fine in 0 to 268435456-1.
#
#		real    4m50.528s
#		user    4m50.730s
#		sys     0m0.021s
#

File: porting-from-algo-c/examples/sqrt.lua
--
--	from src/sqrt.c
--
--	double mysqrt(double)		to	fSqrt
--	fSqrt				to	fSqrtM (depends on lbc(*))
--
--	from src/isqrt.c
--
--	unsigned isqrt(unsigned)	to	iSqrt
--
--	*) bc library for Lua 5.4 / Jul 2018 / based on GNU bc-1.07
--	(lbc-101; see https://web.tecgraf.puc-rio.br/~lhf/ftp/lua/#lbc)
--

local M = require 'nthroot'

local fSqrt, fSqrtM, iSqrt = M.fSqrt, M.fSqrtM, M.iSqrt

local function _t_iSqrt(l, r)
	for i=l,r-1 do
		local t = iSqrt(i)
		if t * t > i or (t+1) * (t+1) <= i then
			error(("ERROR: (i, t) is (%d, %d)"):format(i, t))
		end
	end
	print(("iSqrt() seems to be fine in %d to %d-1."):format(l, r))
end

do
	for i=0,20 do
		print(
			("math.sqrt(%d)\t%.20f\nfSqrt(%d)\t%.20f (%q)")
				:format(
					i,
					math.sqrt(i),
					i,
					fSqrt(i),
					math.sqrt(i) == fSqrt(i)
				)
		)
		if fSqrtM ~= nil then
			print(("fSqrtM(%d,%d)\t%s"):format(i, 50, fSqrtM(i, 50)))
		end
	end

	print("--")

	-- _t_iSqrt(0, 1 << 32) -- Maybe this will take a lot of time.
	_t_iSqrt(0, 1 << 16)
end
